### PR TITLE
cmd/openshift-install/migrate: sub command for migrating azure legacy private zones

### PR DESCRIPTION
**Pull request by jhixson74**
Azure has stopped supporting the previous generation of private DNS zones. This
adds migration support to the installer to move to the new style of private DNS
zones.

https://issues.redhat.com/browse/CORS-1291

### CODE REVIEW

File: pkg/migrate/azure/dns.go [hunk 1]
```diff
@@ -0,0 +1,1335 @@
+package azure
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/dns.go]: i would like this package to be `pkg/migrate/azure/privatedns`

File: pkg/migrate/azure/dns.go [hunk 2]
```diff
@@ -0,0 +1,1335 @@
...
+	return &legacyDNSClient{resourceGroup, public, zonesClient, recordsetsClient}
+}
+
+// safe pointer handling for int32
+func safeInt32(n *int32) int32 {
+	if n == nil {
+		return 0
+	}
+	return *n
+}
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/dns.go]: i think you can use https://godoc.org/github.com/Azure/go-autorest/autorest/to

File: pkg/migrate/azure/dns.go [hunk 3]
```diff
@@ -0,0 +1,1335 @@
...
+	return *s
+}
+
+// safe pointer handling for array of string
+func safeStringArray(s *[]string) string {
+	if s == nil {
+		return ""
+	}
+	return (*s)[0]
+}
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/dns.go]: i hope all of it can be replaced by https://godoc.org/github.com/Azure/go-autorest/autorest/to

File: pkg/migrate/azure/dns.go [hunk 4]
```diff
@@ -0,0 +1,1335 @@
...
+		return ""
+	}
+
+	parts2 := strings.Split(parts1[1], "/providers")
+	if len(parts2) < 2 {
+		return ""
+	}
+
+	return parts2[0]
+}
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/dns.go]: can you add a unit test for this..?
also this needs a little bit more documentation on the function.

File: pkg/migrate/azure/dns.go [hunk 5]
```diff
@@ -0,0 +1,1335 @@
...
+	if err != nil {
+		return err
+	}
+
+	privateDNSZoneInfo, err := deserializePrivateZone(szone)
+	if err != nil {
+		return err
+	}
+
+	doListPrivateZoneDeserialized(privateDNSZoneInfo)
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/dns.go]: can you explain a bit why there is a requirement for deepcopy? and why are we serializing de serializing? also why aren't we using json in that case, instead of `gob` ?

File: pkg/migrate/azure/dns.go [hunk 6]
```diff
@@ -0,0 +1,1335 @@
...
+		}
+		if rs.PtrRecords != nil && len(*rs.PtrRecords) > 0 {
+			showPrivatePtrRecords(rs.PtrRecords, zoneName, *rs.Fqdn, ttl)
+		}
+	}
+
+}
+
+// DoListPrivateZone Retrives private zone info and prints it out
+func DoListPrivateZone(resourceGroup string, privateZone string) error {
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/dns.go]: also this `Do*` is really not necessary, `ListPrivateZone` seems like good enough.

File: pkg/migrate/azure/dns.go (hunk 3 above)
Comment (abhinavdahiya) [reply pkg/migrate/azure/dns.go]: if that doesn't cut it use the `k8s.io/utils/pointer`

File: cmd/openshift-install/migrate.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	if azureMigrateListOpts.legacyZone != "" {
+		return azmigrate.DoListLegacyZone(azureMigrateListOpts.resourceGroup, azureMigrateListOpts.legacyZone, azureMigrateListOpts.public)
+	}
+
+	return azmigrate.DoListPrivateZone(azureMigrateListOpts.resourceGroup, azureMigrateListOpts.privateZone)
+}
+
+func newMigrateAzureDNSListCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "list",
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate.go]: why is this required, can't users just use az to list a zone?

File: cmd/openshift-install/migrate.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
+package main
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate.go]: i would like the structure to be

```
migrate azure privatedns <eligible | migrate>
```

File: cmd/openshift-install/migrate.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+	if azureMigrateTestOpts.zone == "" || azureMigrateTestOpts.resourceGroup == "" {
+		return cmd.Help()
+	}
+
+	return azmigrate.DoTest(azureMigrateTestOpts.zone, azureMigrateTestOpts.resourceGroup)
+}
+
+func newMigrateAzureDNSTestCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "test",
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate.go]: again, this seems like something that shouldn't be added to the cli

File: cmd/openshift-install/migrate.go (hunk 8 above)
Comment (abhinavdahiya) [reply cmd/openshift-install/migrate.go]: and hopefully, the package structure looks like
```
openshift-install/
---- migrate
-------- eligible.go
-------- migrate.go
```

File: cmd/openshift-install/migrate.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+)
+
+func runMigrateAzureDNSEligibleCmd(cmd *cobra.Command, args []string) error {
+	return azmigrate.DoEligible(azureMigrateEligibleOpts.public)
+}
+
+func newMigrateAzureDNSEligibleCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "eligible",
+		Short: "Show legacy zones that are eligible to be migrated",
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate.go]: would like to see long description also added for this.

File: pkg/migrate/azure/dns.go (hunk 1 above)
Comment (jhixson74) [reply pkg/migrate/azure/dns.go]: Fixed.

File: pkg/migrate/azure/dns.go (hunk 2 above)
Comment (jhixson74) [reply pkg/migrate/azure/dns.go]: Nice library, glad you pointed me to it. I've updated the code to use it!

File: pkg/migrate/azure/dns.go (hunk 3 above)
Comment (jhixson74) [reply pkg/migrate/azure/dns.go]: Done!

File: pkg/migrate/azure/dns.go (hunk 4 above)
Comment (jhixson74) [reply pkg/migrate/azure/dns.go]: I've updated this function a bit, and also commented it a bit more.

File: pkg/migrate/azure/dns.go (hunk 5 above)
Comment (jhixson74) [reply pkg/migrate/azure/dns.go]: I've commented the code here for the reasons why it was done this way.

File: cmd/openshift-install/migrate.go (hunk 7 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate.go]: I've removed the leftover cruft.

File: pkg/migrate/azure/dns.go (hunk 6 above)
Comment (jhixson74) [reply pkg/migrate/azure/dns.go]: I've renamed everything to not have Do*.

File: cmd/openshift-install/migrate.go (hunk 8 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate.go]: I've updated the directory structure to this. I've added an extra azure directory where the command files are under. If you don't like that I can remove it and put it this way.

File: cmd/openshift-install/migrate.go (hunk 9 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate.go]: Nuked!

File: cmd/openshift-install/migrate.go (hunk 10 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate.go]: I've added long descriptions.They aren't a whole lot longer, but a bit more descriptive.

**Comment by jhixson74** [comment]
Documentation still needs to be updated.

**Comment by jhixson74** [comment]
After removing the list and test commands, I can probably remove several functions from this.

File: cmd/openshift-install/migrate/azure/migrate.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+	cmd := &cobra.Command{
+		Use:   "migrate",
+		Short: "Migrate a zone",
+		Long:  "This will migrate a legacy private zone to a new style private zone.",
+		RunE:  runMigrateAzurePrivateDNSMigrateCmd,
+	}
+
+	cmd.PersistentFlags().StringVar(&azureMigrateOpts.zone, "zone", "", "The zone to migrate")
+	cmd.PersistentFlags().StringVar(&azureMigrateOpts.oldResourceGroup, "oldResourceGroup", "", "The resource group of the legacy zone")
+	cmd.PersistentFlags().StringVar(&azureMigrateOpts.newResourceGroup, "newResourceGroup", "", "The resource group to create the private zone in")
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate/azure/migrate.go]: this doesn't seem like it's going to useful for openshift clusters.

File: cmd/openshift-install/migrate/azure/migrate.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+	cmd.PersistentFlags().StringVar(&azureMigrateOpts.virtualNetwork, "virtualNetwork", "", "The virtual network to create the private zone in")
+	cmd.PersistentFlags().StringVar(&azureMigrateOpts.vnetResourceGroup, "vnetResourceGroup", "", "The resource group the virtual network is in")
+	cmd.PersistentFlags().BoolVar(&azureMigrateOpts.link, "link", false, "Link the newly created private zone to the virtual network")
+	cmd.PersistentFlags().BoolVar(&azureMigrateOpts.public, "public", false, "Migrate a legacy public zone to a private zone")
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate/azure/migrate.go]: when will this be useful in openshift context.

File: cmd/openshift-install/migrate/azure/migrate.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+func NewMigrateAzurePrivateDNSMigrateCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "migrate",
+		Short: "Migrate a zone",
+		Long:  "This will migrate a legacy private zone to a new style private zone.",
+		RunE:  runMigrateAzurePrivateDNSMigrateCmd,
+	}
+
+	cmd.PersistentFlags().StringVar(&azureMigrateOpts.zone, "zone", "", "The zone to migrate")
+	cmd.PersistentFlags().StringVar(&azureMigrateOpts.oldResourceGroup, "oldResourceGroup", "", "The resource group of the legacy zone")
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate/azure/migrate.go]: I think the fails shouldn't be in camel case, rather `-` separated

File: pkg/migrate/azure/dns.go (hunk 5 above)
Comment (abhinavdahiya) [reply pkg/migrate/azure/dns.go]: @jhixson74 I can't see where you added the comment, can you point me to it in new changset?

Also can you summarize why here too if possible?

File: cmd/openshift-install/migrate/azure/migrate.go (hunk 12 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate/azure/migrate.go]: Removed.

File: cmd/openshift-install/migrate/azure/migrate.go (hunk 13 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate/azure/migrate.go]: Updated.

File: cmd/openshift-install/migrate/azure/migrate.go (hunk 11 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate/azure/migrate.go]: Can we guarantee that the legacy style private domain and the new style private domain will always be in the same resource group? Or can we make that a requirement? These options exist so that they can be different. If a new resource group is not specified, the old resource group is the default. Perhaps that could be made more clear.

File: pkg/migrate/azure/dns.go (hunk 5 above)
Comment (jhixson74) [reply pkg/migrate/azure/dns.go]: It got nuked when this section of code got nuked ;-) Here: https://github.com/openshift/installer/commit/50477870a9fa434e354b6bb39f4d8bf53f932e6f#diff-4727bf8fc830a99ae17aaee2bf9f0021R1084

Comment (abhinavdahiya) [reply pkg/migrate/azure/dns.go]:
```
	// Serialize/Deserialize to make deep copy
	// It is necessary to make a deep copy since there are several
	// references in various structures that go out of scope, I assume they
	// are garbage collected. The necessity to use gob is that most of
	// these structures have JSON marshal/unmarshal methods that do not
	// include necessary properties that are needed here.
```

just for discussion sake.

go shouldn't drop memory that's still in use. so i don't think this serialize/serialize is necessary.. 
can you provide exact example of the error you are seeing or drop these please.

File: cmd/openshift-install/migrate.go [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+	for _, subCmd := range []*cobra.Command{
+		azmigrate.NewMigrateAzurePrivateDNSEligibleCmd(),
+		azmigrate.NewMigrateAzurePrivateDNSMigrateCmd(),
+	} {
+		cmd.AddCommand(subCmd)
+	}
+
+	return cmd
+}
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate.go]: Can we move these to the corresponding packages.

File: cmd/openshift-install/migrate/azure/eligible.go [hunk 15]
```diff
@@ -0,0 +1,23 @@
+package azure
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate/azure/eligible.go]: the file name is `azure/eligible.go` but it only have `privatedns eligible` command, can we rename this file to reflect what it has.

File: cmd/openshift-install/migrate/azure/migrate.go [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+		virtualNetwork    string
+		vnetResourceGroup string
+		link              bool
+	}
+)
+
+func runMigrateAzurePrivateDNSMigrateCmd(cmd *cobra.Command, args []string) error {
+	if (azureMigrateOpts.zone == "" || azureMigrateOpts.oldResourceGroup == "") || (azureMigrateOpts.link == true && azureMigrateOpts.virtualNetwork == "") || (azureMigrateOpts.virtualNetwork != "" && azureMigrateOpts.vnetResourceGroup == "") {
+		return cmd.Help()
+	}
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate/azure/migrate.go]: can we return more helpful error messages instead of just the cmd help.

File: cmd/openshift-install/migrate/azure/migrate.go (hunk 11 above)
Comment (abhinavdahiya) [reply cmd/openshift-install/migrate/azure/migrate.go]: in the context of the openshift clusters, both the private dns and dns private resources will be in same resource group.

the vnet can be different resource group so having a flag for that is nice..., but defaulting it to the cluster resource group will be helpful.

File: cmd/openshift-install/migrate/azure/eligible.go [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+func runMigrateAzurePrivateDNSEligibleCmd(cmd *cobra.Command, args []string) error {
+	return azmigrate.Eligible()
+}
+
+// NewMigrateAzurePrivateDNSEligibleCmd adds the eligble command to openshift-install
+func NewMigrateAzurePrivateDNSEligibleCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "eligible",
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate/azure/eligible.go]: so this currently is `openshift-install migrate azure privatedns eligible` ?

File: cmd/openshift-install/migrate/azure/migrate.go [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+	return azmigrate.Migrate(azureMigrateOpts.oldResourceGroup, azureMigrateOpts.newResourceGroup, azureMigrateOpts.zone, azureMigrateOpts.virtualNetwork, azureMigrateOpts.vnetResourceGroup, azureMigrateOpts.link)
+}
+
+// NewMigrateAzurePrivateDNSMigrateCmd adds the migrate command to openshift-install
+func NewMigrateAzurePrivateDNSMigrateCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "migrate",
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate/azure/migrate.go]: so this currently is `openshift-install migrate azure privatedns migrate` ? now that is sounding really bad. can we improve this a little bit...

File: docs/user/azure/migrate.md [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+Verify this is in fact, a legacy zone:
+
+![azure 4](images/azure-migrate-dns-4.png)
+
+Now, list zones that are eligible for migrating:
+
+```
+[jhixson@redlap test]$ ./openshift-install migrate azure privatedns eligible
+INFO[0000] Credentials loaded from file "/home/jhixson/.azure/osServicePrincipal.json" 
+INFO[0002] legacy zone=testmigrate.installer.azure.devcluster.openshift.com resourceGroup=jhixson_dns_migrate_rg
```
Comment (abhinavdahiya) [inline docs/user/azure/migrate.md]: why is this not using the logging format like the rest of the install commands? Can we fix this?

Comment (abhinavdahiya) [reply docs/user/azure/migrate.md]: esp `INFO[0002]`

File: pkg/migrate/azure/privatedns.go [hunk 20]
```diff
@@ -0,0 +1,744 @@
...
+	recordsets []*azdns.RecordSet
+}
+
+type legacyDNSClient struct {
+	resourceGroup    string
+	zonesClient      azdns.ZonesClient
+	recordsetsClient azdns.RecordSetsClient
+}
+
+func newLegacyDNSClient(session *azconfig.Session, resourceGroup string) *legacyDNSClient {
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/privatedns.go]: this doesn't use resource group, so maybe drop it.

File: pkg/migrate/azure/privatedns.go [hunk 21]
```diff
@@ -0,0 +1,744 @@
...
+					}
+					privateDNSZoneSerialized.recordsets = append(privateDNSZoneSerialized.recordsets, srs)
+				}
+			}
+			privateDNSZonesSerialized = append(privateDNSZonesSerialized, &privateDNSZoneSerialized)
+		}
+	}
+
+	return privateDNSZonesSerialized, nil
+}
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/privatedns.go]: 1. this function is listing all the zones and then listing all recordsets for all these zones, when actually we shouldn't need to list records for zones we are not going to migrate.. so we should be able to reduce these API calls.

So i think this function should be reduce to listing the zones only. the returned value should provide
{ID, resourceGroupID, type}

2. I see no reason for all the serialize deserialize here, so we should drop them.

File: pkg/migrate/azure/privatedns.go [hunk 22]
```diff
@@ -0,0 +1,744 @@
...
+	for _, szone := range szones {
+		legacyDNSZoneInfo, err := deserializeLegacyZone(szone)
+		if err != nil {
+			return err
+		}
+
+		eligibleZone := legacyDNSZoneInfo.zone
+		resourceGroup := idToResourceGroup(*eligibleZone.ID)
+
+		logrus.Printf("legacy zone=%s resourceGroup=%s\n", *eligibleZone.Name, resourceGroup)
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/privatedns.go]: use logrus's fiedLogger to print these fields?

also why are you using Printf instead of `Debug(f)` or `Info(f)` ?

File: pkg/migrate/azure/privatedns.go [hunk 23]
```diff
@@ -0,0 +1,744 @@
...
+				txtRecords = append(txtRecords, *txtRecord)
+			}
+		}
+	}
+
+	return &txtRecords
+}
+
+// Transforms a legacy zone to a private zone
+func (client *privateDNSClient) MigrateLegacyZone(legacyDNSZoneInfo *legacyDNSZoneInfo, link bool) error {
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/privatedns.go]: why is this function public?

File: pkg/migrate/azure/privatedns.go [hunk 24]
```diff
@@ -0,0 +1,744 @@
...
+					}
+					legacyDNSZoneSerialized.recordsets = append(legacyDNSZoneSerialized.recordsets, srs)
+				}
+			}
+			legacyDNSZonesSerialized = append(legacyDNSZonesSerialized, &legacyDNSZoneSerialized)
+		}
+	}
+
+	return legacyDNSZonesSerialized, nil
+}
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/privatedns.go]: Same as the other GetZonesSerialized

File: pkg/migrate/azure/privatedns.go (hunk 23 above)
Comment (abhinavdahiya) [reply pkg/migrate/azure/privatedns.go]: also why are we calling Migrate on the new zone type client? isn't that reverse? we should be calling migrate on the legacy zone client??

File: pkg/migrate/azure/privatedns.go [hunk 25]
```diff
@@ -0,0 +1,744 @@
...
+
+	location := "global"
+	virtualNetworkID := *virtualNetwork.ID
+	virtualNetworkLinkName := fmt.Sprintf("%s-network-link", strings.Replace(client.vnetResourceGroup, "-rg", "", 1))
+
+	// pass a flag for this?
+	registrationEnabled := false
+
+	virtualNetworkLink := azprivatedns.VirtualNetworkLink{
+		Location: &location,
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/privatedns.go]: maybe use the `autorest/to` here

File: pkg/migrate/azure/privatedns.go [hunk 26]
```diff
@@ -0,0 +1,744 @@
...
+		VirtualNetworkLinkProperties: &azprivatedns.VirtualNetworkLinkProperties{
+			VirtualNetwork: &azprivatedns.SubResource{
+				ID: &virtualNetworkID,
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/privatedns.go]: same as https://github.com/openshift/installer/pull/3343/files#r399574726

File: pkg/migrate/azure/dns.go (hunk 5 above)
Comment (jhixson74) [reply pkg/migrate/azure/dns.go]: I can provide an example, but it's no longer necessary since I have nuked most of this code from orbit ;-).

File: cmd/openshift-install/migrate/azure/migrate.go (hunk 11 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate/azure/migrate.go]: I've removed the options for new/old resource group. Now, just the resource group is required.

File: cmd/openshift-install/migrate.go (hunk 14 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate.go]: They've been moved.

File: cmd/openshift-install/migrate/azure/migrate.go (hunk 16 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate/azure/migrate.go]: I've updated the error messages here.

File: cmd/openshift-install/migrate/azure/eligible.go (hunk 17 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate/azure/eligible.go]: Now it is azure-privatedns-migrate and azure-privatedns-eligible, which I am not crazy about either, but simpler.

File: docs/user/azure/migrate.md (hunk 19 above)
Comment (jhixson74) [reply docs/user/azure/migrate.md]: I've updated the doc to reflect the changes.

File: pkg/migrate/azure/privatedns.go (hunk 20 above)
Comment (jhixson74) [reply pkg/migrate/azure/privatedns.go:31]: the resource group is used/needed, so it's still here.

File: pkg/migrate/azure/privatedns.go (hunk 21 above)
Comment (jhixson74) [reply pkg/migrate/azure/privatedns.go]: As stated earlier, there was indeed an issue, and I can demonstrate it. While my curiosity wants to know what it is, it is no longer relevant to this code, so marking this as resolved.

File: pkg/migrate/azure/privatedns.go (hunk 24 above)
Comment (jhixson74) [reply pkg/migrate/azure/privatedns.go]: Nuked.

File: pkg/migrate/azure/privatedns.go (hunk 25 above)
Comment (jhixson74) [reply pkg/migrate/azure/privatedns.go]: Updated to use autorest/to.

File: pkg/migrate/azure/privatedns.go (hunk 26 above)
Comment (jhixson74) [reply pkg/migrate/azure/privatedns.go]: Updated to use autorest/to.

File: pkg/migrate/azure/privatedns.go (hunk 22 above)
Comment (jhixson74) [reply pkg/migrate/azure/privatedns.go]: No particular reason other than my personal preference. Printf() is same as Info(). I have updated it to use Info() to be consistent with the rest of the installer code, however.

File: pkg/migrate/azure/privatedns.go (hunk 23 above)
Comment (jhixson74) [reply pkg/migrate/azure/privatedns.go]: I've made the function private.

File: cmd/openshift-install/migrate/azure/migrate.go [hunk 27]
```diff
@@ -0,0 +1,53 @@
...
+		resourceGroup     string
+		virtualNetwork    string
+		vnetResourceGroup string
+		link              bool
+	}
+)
+
+func runMigrateAzurePrivateDNSMigrateCmd(cmd *cobra.Command, args []string) error {
+	if azureMigrateOpts.zone == "" {
+		logrus.Fatal("zone is a required argument")
```
Comment (abhinavdahiya) [inline cmd/openshift-install/migrate/azure/migrate.go:22]: you can return an error here instead of fatal error..

File: pkg/migrate/azure/privatedns.go [hunk 28]
```diff
@@ -0,0 +1,529 @@
...
+	if err != nil {
+		return err
+	}
+
+	// Read back the newly created zone to verify creation
+	_, err = client.zonesClient.Get(ctx, client.resourceGroup, *privateZone.Name)
+	if err != nil {
+		return err
+	}
+	logrus.Infof("ok.\n")
```
Comment (abhinavdahiya) [inline pkg/migrate/azure/privatedns.go:419]: i don't think you need `\n` with infof

Comment (abhinavdahiya) [reply pkg/migrate/azure/privatedns.go:419]: and this can be info instead of infof

**Comment by abhinavdahiya** [comment]
/approve

I think you are missing a `go mod tidy && go mod vendor` to update the vendor..  https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/3343/pull-ci-openshift-installer-master-verify-vendor/4600 FAILED.

File: cmd/openshift-install/migrate/azure/migrate.go (hunk 27 above)
Comment (jhixson74) [reply cmd/openshift-install/migrate/azure/migrate.go]: Fixed.

File: pkg/migrate/azure/privatedns.go (hunk 28 above)
Comment (jhixson74) [reply pkg/migrate/azure/privatedns.go]: Fixed.

File: cmd/openshift-install/main.go [hunk 29]
```diff
@@ -54,6 +55,8 @@ func installerMain() {
 		newVersionCmd(),
 		newGraphCmd(),
 		newCompletionCmd(),
+		azure.NewMigrateAzurePrivateDNSEligibleCmd(),
+		azure.NewMigrateAzurePrivateDNSMigrateCmd(),
```
Comment (abhinavdahiya) [inline cmd/openshift-install/main.go:59]: this puts these at the same level as the `create, destroy`.. that is not correct.
```
openshift-install migrate azure-privatedns-eligible
openshift-install migrate azure-privatedns
```

i'm sorry for not looking at this before, but i don't think we can keep these at root same as `create, destroy`.

Comment (jhixson74) [reply cmd/openshift-install/main.go]: Fixed!
